### PR TITLE
A big options clean-up

### PR DESCRIPTION
**Pull request by nnethercote**
Lots of improvements here.

r? @Centril

### CODE REVIEW

**Comment by nnethercote** [comment]
I recommend reviewing this one commit at a time.

File: src/librustc_session/options.rs [hunk 1]
```diff
@@ -875,13 +875,16 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         `hir,typed` (HIR with types for each node),
         `hir-tree` (dump the raw HIR),
         `mir` (the MIR), or `mir-cfg` (graphviz formatted MIR)"),
-    run_dsymutil: Option<bool> = (None, parse_opt_bool, [TRACKED],
+    // The default historical behavior was to always run dsymutil, so we're
+    // preserving that temporarily, but we're likely to switch the default
+    // soon.
```
Comment (fbstj) [inline src/librustc_session/options.rs:887]: does this deserve a FIXME or a tracking issue?

File: src/librustc_session/options.rs [hunk 2]
```diff
@@ -480,18 +481,16 @@ macro_rules! options {
...
-                }
+            match v {
+                Some(s) => {
+                    match s {
+                        "0" => { *slot = 0; true }
+                        "1" => { *slot = 1; true }
+                        "2" => { *slot = 2; true }
+                        _ => { false }
+                    }
+                },
```
Comment (Centril) [inline src/librustc_session/options.rs:472]:
```suggestion
                Some("0") => { *slot = 0; true }
                Some("1") => { *slot = 1; true }
                Some("2") => { *slot = 2; true }
                Some(_) => false,
```

File: src/librustc_session/options.rs [hunk 3]
```diff
@@ -622,21 +622,21 @@ options! {CodegenOptions, CodegenSetter, basic_codegen_options,
...
-        "save all temporary output files during compilation"),
+        "save all temporary output files during compilation (default: no)"),
     rpath: bool = (false, parse_bool, [UNTRACKED],
-        "set rpath values in libs/exes"),
+        "set rpath values in libs/exes (default: no)"),
     overflow_checks: Option<bool> = (None, parse_opt_bool, [TRACKED],
         "use overflow checks for integer arithmetic"),
     no_prepopulate_passes: bool = (false, parse_bool, [TRACKED],
-        "don't pre-populate the pass manager with a list of passes"),
+        "give an empty list of passes with the pass manager (default: no)"),
```
Comment (Centril) [inline src/librustc_session/options.rs:631]:
```suggestion
        "give an empty list of passes to the pass manager (default: no)"),
```

File: src/doc/rustc/src/codegen-options/index.md [hunk 4]
```diff
@@ -275,24 +291,26 @@ This option is deprecated and does nothing.
...
-* `3`: all optimizations
-* `s`: optimize for binary size
-* `z`: optimize for binary size, but also turn off loop vectorization.
+* `0`: no optimizations, also turns on
+  [`cfg(debug_assertions)`](#debug-assertions) (the default).
+* `1`: basic optimizations.
+* `2`: some optimizations.
+* `3`: all optimizations.
+* `s`: optimize for binary size.
+* `z`: optimize for binary size, but also turn off loop vectorization..
```
Comment (Centril) [inline src/doc/rustc/src/codegen-options/index.md:313]:
```suggestion
* `z`: optimize for binary size, but also turn off loop vectorization.
```

**Review by Centril — COMMENTED** [review]
This looks nice. Here are some things I spotted during review.
However, I'm not too familiar with this part of the codebase, so I'll leave approval to @petrochenkov.

**Comment by petrochenkov** [comment]
This is great, thank you.

After our discussion about `-Cembed-bitcode` I had the same thought about converting *all* the flag options (old `parse_bool`) into boolean options (old `parse_opt_bool`).

**Comment by petrochenkov** [comment]
I have some reservations regarding all the `no-*` options which look pretty bad with the new `yes/no` values due to possible double negation.
With the new scheme I think we need to convert them from `no-foo=no/yes` to `foo=yes/no`.

For unstable flag options `-Z no-foo` we can
- Add `-Z foo=yes/no`
- Remove the existing `-Z no-foo` (possibly later in a separate PR)

For stable flag options `-C no-foo` we can
- Add `-C foo=yes/no`
- Keep the existing `-C no-foo` for compatibility, without support for `yes/no` (by introducing `parse_flag` equivalent to the old `parse_bool`)

**Comment by nnethercote** [comment]
> (Or we can just avoid touching the no-* options in this PR at all.)

I would strongly prefer that.

I agree that `--no-foo` flags are weird and it would be nice to get rid of them. Having said that, I think that addition/removal/renaming of flags is significant scope creep for this PR, which is about cleaning things up and making things consistent without any significant changes in behaviour.
(Also note that someone writing `--no-foo=no` is extremely unlikely, because (a) it has been impossible prior to now, and (b) it has no effect unless preceded by `--no-foo`.)

**Comment by petrochenkov** [comment]
>I would strongly prefer that.

Ok.
Could you reintroduce `parse_flag` then and use it for the `no-*` options.

**Comment by nnethercote** [comment]
> Could you reintroduce parse_flag then and use it for the no-* options.

Advantages of this:
- Double-negatives (e.g. `--no-foo=no`) are not possible.

Disadvantages of this:
- `--no-foo` options would go back to being sticky, i.e. not overridable.
- We would lose the new consistency that all options can take values.
- We would require an extra `parse_*` function.
- We would have to remove the "Make option type descriptions non-optional" commit, which was a nice simplification.

In my opinion, the disadvantages outweigh the advantages, particularly given that the number of `--no-foo` options is likely to drop in the future.

**Comment by petrochenkov** [comment]
@nnethercote 
Looks like we misinterpreted each other then?

In "avoid touching the no-* options in this PR at all" I meant not touching them compared to the status quo (before this PR), not compared to this PR as is.
You were also talking about "making things consistent without any significant changes" which I also interpreted as support for the status quo.

**Comment by petrochenkov** [comment]
>Double-negatives (e.g. `--no-foo=no`) are not possible.

`--no-foo=yes`, which is equally confusing, also becomes impossible.

>`--no-foo` options would go back to being sticky, i.e. not overridable.

I'll fix this and introduce the corresponding `--foo` options once the option field redirection from https://github.com/rust-lang/rust/pull/70665 lands. (That PR will be less about cleanup, and more about significant changes.)

>We would lose the new consistency that all options can take values.

Temporarily, see the previous comment.

>We would require an extra `parse_*` function.

Certainly not a big deal.

>We would have to remove the "Make option type descriptions non-optional" commit

Why can't we come up with some description for `parse_flag` and keep it? It's indeed a nice simplification.

**Comment by nnethercote** [comment]
Sorry for the confusion. I like the PR code as is. It deliberately introduces the minor functional change that all boolean options, including `--no-foo` ones, can take a value. I call this "minor" because it doesn't require any rustc user to change how they use the compiler. (Though they can, if they want to.)

In contrast, I consider renaming flags or removing flags "non-minor" because rustc users might have to change how they use the compiler.

**Comment by nnethercote** [comment]
Hmm, `-C no-redzone` complicates things. It's an existing `no` option that already takes a value.

A problem with the `parse_flag` suggestion: if we end up with `-C no-foo` and `-C foo=<bool>`, then any use site has to check for both `cg.no_foo` and `cg.foo`, which is error-prone.

Here's another possibility that avoids that problem, and also gets us away from the use of `no` more quickly.
- Switch all the `--no-foo` options to `--foo=<bool>`...
- ...but give them a marking that allows `--no-foo` to be allowed as a synonym for `--foo=no`.
- In the documentation, use the `--foo=` form, but mention that `--no-foo` is a deprecated alternative.
- Later on (though perhaps quite soon), remove the marking for all the current `-Z no-foo` options.
- We'll probably need to keep the marking for the current `-C no-foo` options for some time, or forever.

**Comment by eddyb** [comment]
cc @rust-lang/compiler

**Comment by petrochenkov** [comment]
@nnethercote 
>A problem with the `parse_flag` suggestion: if we end up with `-C no-foo` and `-C foo=<bool>`, then any use site has to check for both `cg.no_foo` and `cg.foo`, which is error-prone.

I wanted to use the same method that is used in https://github.com/rust-lang/rust/pull/70665 for merging `-C link-arg` into `-C link-args` to merge `-C no-foo` into `-C foo` and avoid the mentioned error prone situation.
The `macro_rules! redirect_field` table serves as the marking in that case, but perhaps it can be integrated with the `options!` DSL in a nicer way.

File: src/librustc_session/options.rs (hunk 2 above)
Comment (nnethercote) [reply src/librustc_session/options.rs:472]: Nice suggestion. And I can combine the `Some("2")` and `None` arms.

File: src/librustc_session/options.rs (hunk 1 above)
Comment (nnethercote) [reply src/librustc_session/options.rs:887]: Perhaps? The comment was pre-existing, I just moved it, so I don't really know.

**Comment by nnethercote** [comment]
I added a commit with a `parse_no_flag` function that reverts the changes to the `no-` flags. This commit is highly prone to conflicts and that seemed like the fastest way to get it landed.

**Comment by nikomatsakis** [comment]
Can someone summarize in what way this is changing the "stable surface API" of the compiler? I saw some back and forth above regarding "double negatives" but I wasn't quite sure where it landed.

**Comment by petrochenkov** [comment]
@nikomatsakis 
I'll write a summary a bit later today (and start the FCP).

**Comment by petrochenkov** [comment]
This PR turns most of the "flag" command line options `-C foo` into "boolean" command line options `-C foo=yes/no`.

---
The benefit of "boolean" options is that they can override each other on the command line, so hierarchical build systems can set one option value for most of the code and then override it for some specific subsets of code by appending options to the command line.

For example, compiletest sets `-Z deduplicate-diagnostics=no` by default for UI tests, but individual tests can opt-out of this by using `// compile-flags: -Z deduplicate-diagnostics=yes`.

For another example, the author of #58211 tried to introduce a single common lint list for multiple projects, then perhaps override some of the lints for individual projects (that issue resulted in lint options becoming overridable).

Unfortunately, rustc options cannot be used in hierarchical way with `.cargo/config`s right now, because cargo concatenates them in the opposite order (from more specific to less specific), but maybe that can be fixed (cc @ehuss).

---

The transformations from "flag" to "boolean" is applied to `-C` and `-Z` options.

Options starting with `no-*` (e.g. `-C no-foo`) are not changed because `-C no-foo=yes` and `-C no-foo=no` look bad, there are plans to change them to `-C foo=yes/no` in a separate PR, and remove or deprecate the `no-*` versions (for unstable or stable options respectively).

@rfcbot fcp merge

**Comment by nnethercote** [comment]
> This PR turns most of the "flag" command line options -C foo into "boolean" command line options -C foo=yes/no.

It's worth emphasizing that you can still use `-C foo` without a value, and it is equivalent to `-C foo=yes`. So all existing valid command lines will remain valid, and some currently invalid command lines will become valid.

**Comment by nnethercote** [comment]
One more approval is needed, from one of @estebank, @matthewjasper, or @pnkfelix.

**Comment by petrochenkov** [comment]
r=me once the FCP is completed

**Pull request merged**